File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.java
class Solution {
    public String evaluate(String s, List<List<String>> knowledge) {

        int n = s.length();

        HashMap<String, String> mp = new HashMap<>();

        for (List<String> vec : knowledge) {
            mp.put(vec.get(0), vec.get(1));
        }

        StringBuilder result = new StringBuilder();

        int i = 0;

        while (i < n) {

            if (Character.isLetter(s.charAt(i))) {
                result.append(s.charAt(i));

            } else { // '('
                i++;

                StringBuilder temp = new StringBuilder();

                while (i < n && s.charAt(i) != ')') {
                    temp.append(s.charAt(i));
                    i++;
                }

                String key = temp.toString();

                if (mp.containsKey(key)) {
                    result.append(mp.get(key));
                } else {
                    result.append("?");
                }
            }

            i++;
        }

        return result.toString();
    }
}

File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.py
class Solution:
    def evaluate(self, s: str, knowledge: list[list[str]]) -> str:
        mp = {k: v for k, v in knowledge}

        result = []
        i, n = 0, len(s)

        while i < n:
            if s[i] == '(':
                i += 1
                temp = []
                while i < n and s[i] != ')':
                    temp.append(s[i])
                    i += 1
                key = ''.join(temp)
                result.append(mp.get(key, '?'))
            else:
                result.append(s[i])
            i += 1

        return ''.join(result)